# Week-1 Assignment

*Welcome to your first assignment for the SimuTech Winter Project 2022! I hope you are excited to implement and test everything you have learned up until now. There is an interesting set of questions for you to refine your acquired skills as you delve into hands-on coding and deepen your understanding of numpy, pandas, and data visualization libraries.*

# Section 0 : Importing Libraries

*Let's begin by importing numpy, pandas and matplotlib.*

In [2]:
import numpy as np
import pandas as pd
import matplotlib as plt
import urllib

# Section 1 : Playing with Python and Numpy

### Q1. Matrix Multiplication

##### (i) Check if matrix multiplication is valid

In [3]:
def isValid(A,B):
  i,j = A.shape
  k,l = B.shape

  return j==k

##### (ii) Using loops (without using numpy)

In [4]:
def matrix_multiply(A,B):
    
  i,j = A.shape
  k,l = B.shape

  C = result = [[0 for _ in range(l)] for _ in range(i)]

  for a in range(i):
        for b in range(l):
            for c in range(j): 
                C[a][b] += A[a][c] * B[c][b]

  return c

##### (iii) Using numpy

In [5]:
def matrix_multiply_2(A,B):
  return A@B

##### (iv) Testing your code

Run the following cell to check if your functions are working properly.

*Expected output:*
[ [102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600] ]

In [6]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])

B = np.array([
    [13, 14, 15],
    [16, 17, 18],
    [19, 20, 21]
])

if isValid(A,B):
  print(f"Result using loops: {matrix_multiply(A,B)}")
  print(f"Result using numpy: {matrix_multiply_2(A,B)}")
else:
  print(f"Matrix multiplication is not valid")

Result using loops: 2
Result using numpy: [[102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600]]


### Q2. Z-Score Normalisation

Z-score normalization refers to the process of normalizing every value in a dataset such that the mean of all of the values is 0 and the standard deviation is 1.

We use the following formula to perform a z-score normalization on every value in a dataset:

New value = (x – μ) / σ

where:

x: Original value

μ: Mean of data

σ: Standard deviation of data

##### (i) Without using numpy

In [7]:
def mean(x):
  sum =0
  for i in range(len(x)):
    sum += x[i]
  return sum/len(x)  

In [8]:
def standard_deviation(x):
  Mean = mean(x)
  std_sum=0
  for i in range(len(x)):
    std_sum += (x[i]-Mean)**2
  return (std_sum/(len(x)))**(0.5)

In [9]:
def zscore_normalisation(x):
  normalized=[0]*len(x)
  Mean = mean(x)
  Std_dev=standard_deviation(x)
  for i in range(len(x)):
    normalized[i] = (x[i]-Mean)/Std_dev

  return normalized

##### (ii) Using numpy

Numpy has in_built functions for calculating mean and standard deviation

In [10]:
def zscore_normalisation_2(x):
  normalized =normalized=[0]*len(x)
  for i in range(len(x)):
    normalized[i]= (x[i]-np.mean(x))/np.std(x)

  return normalized

##### (iii) Testing your code

Run the following cell to check if your functions are working properly.

*Expected Output:* [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]

In [11]:
x = [4, 7, 7, 15, 32, 47, 63, 89, 102, 131]
print(f"Result without using numpy: {zscore_normalisation(x)}")
print(f"Result using numpy: {zscore_normalisation_2(x)}")

Result without using numpy: [-1.0675326683028088, -0.9974539373420117, -0.9974539373420117, -0.8105773214465528, -0.41346451266870277, -0.06307085786471743, 0.3106823739262003, 0.9180313755864415, 1.2217058764165623, 1.8991336090376005]
Result using numpy: [np.float64(-1.0675326683028088), np.float64(-0.9974539373420117), np.float64(-0.9974539373420117), np.float64(-0.8105773214465528), np.float64(-0.41346451266870277), np.float64(-0.06307085786471743), np.float64(0.3106823739262003), np.float64(0.9180313755864415), np.float64(1.2217058764165623), np.float64(1.8991336090376005)]


### Q3. Sigmoid fn and its derivative

The sigmoid function is a mathematical function that maps any input value to a value between 0 and 1.

It is defined mathematically as s(x) = 1/(1+e^(-x)).

##### (i) Write a fn to implement sigmoid fn

In [12]:
def sigmoidfn(x):
  sigmoid = [0]*len(x)
  for i in range(len(x)):
    sigmoid[i]=1/1+np.exp(-(x[i]))

  return sigmoid

##### (ii) Write a fn to implement derivative of sigmoid fn

In [13]:
def derivative(x):
  dx=[0]*len(x)
  for i in range(len(x)):
    dx[i]=(1/1+np.exp(-(x[i])))**2

  return dx

##### (iii) Test your code

Run the following cell to check if your functions are working properly.

*Expected output:*

x on applying sigmoid activation fn is: [ [0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ] ]

x on applying derivative of sigmoid activation fn is: [ [-1.23379350e-04 -1.04993585e-01 -6.64805667e-03 -2.50000000e-01
  -2.50000000e-01]
 [-9.10221180e-04 -6.64805667e-03 -2.50000000e-01 -2.50000000e-01
  -2.50000000e-01] ]

In [14]:
x = np.array([
    [9,2,5,0,0],
    [7,5,0,0,0]
])
print(f"x on applying sigmoid activation fn is: {sigmoidfn(x)}")
print(f"x on applying derivative of sigmoid activation fn is: {derivative(x)}")

x on applying sigmoid activation fn is: [array([1.00012341, 1.13533528, 1.00673795, 2.        , 2.        ]), array([1.00091188, 1.00673795, 2.        , 2.        , 2.        ])]
x on applying derivative of sigmoid activation fn is: [array([1.00024683, 1.28898621, 1.01352129, 4.        , 4.        ]), array([1.0018246 , 1.01352129, 4.        , 4.        , 4.        ])]


# Section 2: Exploring Pandas

*You have been provided with a dataset which includes information about properties of superheated vapor.*

*The dataset consists of the thermophysical properties: specific volume, specific internal energy, specific enthalpy, specific entropy of superheated vapor.*

*Pressure is in kPa and Temperature in centigrade. In the dataframe 75, 100, 125, etc. are temperatures.*

### Read the csv file


In [15]:
url = "https://raw.githubusercontent.com/TYRO09/VisionCraft-TheWinter-Challenge/main/Week1/superheated_vapor_properties.csv"

filename = "data.csv"
urllib.request.urlretrieve(url, filename)

('data.csv', <http.client.HTTPMessage at 0x1e06ba66690>)

In [16]:
data = pd.read_csv('data.csv')


### Display the shape of data frame


In [17]:
print(data.shape)

(1090, 37)


### Return an array containing names of all the columns

In [18]:
print(data.columns)

Index(['Pressure', 'Property', 'Liq_Sat', 'Vap_Sat', '75', '100', '125', '150',
       '175', '200', '220', '225', '240', '250', '260', '275', '280', '290',
       '300', '320', '325', '340', '350', '360', '375', '380', '400', '425',
       '450', '475', '500', '525', '550', '575', '600', '625', '650'],
      dtype='object')


### Display the number of null values in each column of the dataframe



In [19]:
null_count = data.isnull().sum()
print(null_count)

Pressure       0
Property       2
Liq_Sat        2
Vap_Sat        2
75          1057
100         1017
125          977
150          897
175          769
200          641
220          817
225          801
240          817
250          689
260          769
275          681
280          761
290          977
300          121
320          961
325          273
340          954
350          137
360          954
375          409
380          954
400            2
425          410
450            2
475          410
500            2
525          546
550            2
575          682
600            2
625          954
650            2
dtype: int64


### Create a column which contains the Pressure and Property columns, seperated with 'at' (For eg. V at 1, H at 101.325). Using this print the following:
- Enthalpy at 75 kPa and 573 K
- Entropy at 493 K and 250 kPa



1057


### Find out the column with the highest number of missing values

In [48]:
print(max(null_count))

1057


### What is the average enthalpy of Sat. Liq. at all different pressures in the dataset?

In [33]:
S_data = data[data["Property"] == "S"]

a = S_data.T
columns = list((a.columns))

Mean_entropy = [0]*len(columns)
rep=0
for i in columns:
    Mean_entropy[rep]=a[i].mean
    rep+=1
print(np.array(Mean_entropy))

[<bound method Series.mean of Pressure          1
 Property          S
 Liq_Sat       0.106
 Vap_Sat      8.9767
 75           9.3828
 100          9.5136
 125          9.6365
 150          9.7527
 175          9.8629
 200          9.9679
 220             NaN
 225         10.0681
 240             NaN
 250         10.1641
 260             NaN
 275             NaN
 280             NaN
 290             NaN
 300          10.345
 320             NaN
 325             NaN
 340             NaN
 350         10.5133
 360             NaN
 375             NaN
 380             NaN
 400         10.6711
 425             NaN
 450           10.82
 475             NaN
 500         10.9612
 525             NaN
 550         11.0957
 575             NaN
 600         11.2243
 625             NaN
 650         11.3476
 Name: 3, dtype: object>
 <bound method Series.mean of Pressure         10
 Property          S
 Liq_Sat      0.6493
 Vap_Sat      8.1511
 75           8.3168
 100          8.4486
 125          

### Separate out the V,U,H,S data from the dataset into V_data, U_data, H_data, S_data

In [22]:
V_data = data[data["Property"] == "V"]
U_data = data[data["Property"] == "U"]
H_data = data[data["Property"] == "H"]
S_data = data[data["Property"] == "S"]


# Section 4 : Conculsion




*Congratulations on reaching this point! I hope you had fun solving your first assignment and have also built confidence in applying these libraries. If you are wondering, we will cover more about z-score normalization in Week 2, and the sigmoid function will be used in Week 3. After completing this assignment, you are now prepared to learn about machine learning techniques and implement your own machine learning models.*